bindings/go/src/fdb/range.go

```diff
@@ -170,119 +181,105 @@ func (rr RangeResult) GetSliceOrPanic() []KeyValue {
 
 // Iterator returns a RangeIterator over the key-value pairs satisfying the
 // range specified in the read that returned this RangeResult.
+// Close() must be called on this range iterator to avoid a memory leak.
 func (rr RangeResult) Iterator() *RangeIterator {
 	return &RangeIterator{
 		t:         rr.t,
-		f:         rr.f,
 		sr:        rr.sr,
 		options:   rr.options,
 		iteration: 1,
 		snapshot:  rr.snapshot,
+		more:      true,
 	}
 }
 
 // RangeIterator returns the key-value pairs in the database (as KeyValue
 // objects) satisfying the range specified in a range read. RangeIterator is
 // constructed with the (RangeResult).Iterator method.
 //
-// You must call Advance and get a true result prior to calling Get or MustGet.
+// You must call NextBatch and get a true result.
 //
 // RangeIterator should not be copied or used concurrently from multiple
 // goroutines, but multiple RangeIterators may be constructed from a single
 // RangeResult and used concurrently. RangeIterator should not be returned from
 // a transactional function passed to the Transact method of a Transactor.
 type RangeIterator struct {
-	t         *transaction
-	f         *futureKeyValueArray
-	sr        SelectorRange
-	options   RangeOptions
-	iteration int
-	done      bool
-	more      bool
-	kvs       []KeyValue
-	index     int
-	err       error
-	snapshot  bool
+	t               *transaction
+	sr              SelectorRange
+	options         RangeOptions
+	iteration       int
+	more            bool
+	lastBatchLen    int
+	snapshot        bool
+	prefetchedBatch *futureKeyValueArray
+	prefetchOptions RangeOptions
 }
 
-// Advance attempts to advance the iterator to the next key-value pair. Advance
-// returns true if there are more key-value pairs satisfying the range, or false
-// if the range has been exhausted. You must call this before every call to Get
-// or MustGet.
-func (ri *RangeIterator) Advance() bool {
-	if ri.done {
-		return false
+// NextBatch advances the iterator to the next key-value pairs batch.
+// It returns a slice of key-value pairs satisfying the range, an empty slice
+// if the range has been exhausted, or an error if any of the asynchronous
+// operations associated with this result did not successfully complete.
+// If 'prefetch' is true, future for the next batch will be created on return and
+// released when the RangeIterator is closed.
+func (ri *RangeIterator) NextBatch(prefetch bool) ([]KeyValue, error) {
```

Comment:
I'm fine with the renaming @jzhou77 any thoughts on the renaming?

Reply:
I was looking at Java iterators for inspiration; if you have better ideas happy to change the implementation accordingly.